src/Dapr.Actors/Runtime/ActorManager.cs

```diff
@@ -186,7 +186,13 @@ async Task<object> RequestFunc(Actor actor, CancellationToken ct)
             // Serialize result if it has result (return type was not just Task.)
             if (methodInfo.ReturnType.Name != typeof(Task).Name)
             {
-                await JsonSerializer.SerializeAsync(responseBodyStream, result, result.GetType(), jsonSerializerOptions);
+#if NET7_0_OR_GREATER
+                var returnType = methodInfo.ReturnType.GenericTypeArguments[0];
+                await JsonSerializer.SerializeAsync(responseBodyStream, result, returnType, jsonSerializerOptions);
+#else
+                await JsonSerializer.SerializeAsync<object>(responseBodyStream, result, jsonSerializerOptions); 
```

Comment:
Thanks for reviewing the PR. As I understood things, apart from supporting `null` responses my change did not change the behavior in .NET 6. I understood that in .NET 6 for a non-null response
```C#
await JsonSerializer.SerializeAsync(responseBodyStream, result, result.GetType(), jsonSerializerOptions);
```
and
```C#
await JsonSerializer.SerializeAsync<object>(responseBodyStream, result, jsonSerializerOptions);
```
produce the same json. And as the current version of the code uses the runtime type the json includes properties of derived classes. The E2E tests I added assert that for .NET 6 my change does the same. As such my change should not introduce new privacy or security concerns. But please correct me if I'm wrong.

I'm not sure I understand your comment regarding adding some sort of flag on the options object to support polymorphic (de)serialization in .NET 6. As System.Text.Json in .NET 6 does not support polymorphic (de)serialization out-of-the-box I don't think Dapr needs to support it and a flag is not needed for .NET 6.

System.Text.Json in .NET 7 does support polymorphic (de)serialization out-of-the-box, provided we serialize using the declared type on the interface instead of the runtime type as follows:
 
```C#
var returnType = methodInfo.ReturnType.GenericTypeArguments[0];
await JsonSerializer.SerializeAsync(responseBodyStream, result, returnType, jsonSerializerOptions);
```

I think using the declared type instead of the runtime type should be the default in .NET 7 or later. Again, no flag needed. If programmers require polymorphic (de)serialization then they can use the JsonDerivedTypeAttribute or, as I prefer, write a JsonTypeInfoResolver.

Reply:
Looking at the docs again, you're right, there's no change to the existing .NET 6 usage, so I retract my earlier comments.

This would still represent a breaking change for users of the SDK that rely on the .NET 6 polymorphic behavior when they move to .NET 8 though, right, in that they'd need to decorate their model types with the appropriate attributes?  (I agree that that seems like the best approach.) Given that, we'd need to make sure there was adequate documentation of that change.

Reply:
>>This would still represent a breaking change for users of the SDK that rely on the .NET 6 polymorphic behavior when they move to .NET 8 though, right
Correct, that would be a breaking change. Although one could argue in .NET 6 one has polymorphic serialization, not polymorphic deserialization, but it is exactly the polymorphic serialization that would break when moving to .NET 8.

>>we'd need to make sure there was adequate documentation of that change
As in a release note? 
Or as in Dapr docs somewhere in https://docs.dapr.io/developing-applications/sdks/dotnet/dotnet-actors/ ? 

The Dapr docs have no current mention on polymorphism. It would be good to write about polymorphism in the Dapr docs for both weakly-typed and strongly-typed actors using both DataContractSerializer and json serializer. And also for service invocation.

The E2E tests I added for this change are also the first tests on polymorphism in the .NET SDK if I'm not mistaken. I only added E2E to test polymorphism in a response object for weakly-typed actors, because that was not currently working. It would be good to also add E2E tests for polymorphism in a request object in weakly typed actors. And for both request and ressponse objects in strongly-typed actors using both the DataContractSerializer and the newly added support for json serialization in strongly-typed actors. But I think it is best to add those additional tests in a separate PR so that this PR remains targeted on the fix.

Reply:
#1199 is now #1222 since I had a typo in the DCO.  @RemcoBlok I'd encourage you to type up the documentation in the actors for your changes so it's available immediately in the next release instead of being lost in this PR thread indefinitely. There's a flag you can request they add linking the docs to your PR here so one isn't accepted without the other.

Reply:
Thanks all for your replies. I've been travelling and not had a chance to respond earlier. I have some text for a release note. Is it best to discuss that in the issue I created related to the PR instead of here? The issue template seemed to suggest that the release notes are picked up from the issue, not the PR.

Reply:
On second thoughts, let's discuss the release note here so that we don't have to jump between threads.  

# Release  Note

## Weakly-typed actor should support polymorphic response.

### Problem
The .NET 6 and .NET 8 targets of the Dapr SDK have different serialization behavior of properties of a derived response object on an operation of a weakly-typed actor client.

### Impact
Impacts users upgrading a consumer application to .NET 8 running Dapr <version> or later.

### Root Cause
In the .NET 6 target of the SDK the actor runtime serializes the response object on an operation of a weakly-typed actor client using the runtime type of the response object. In the .NET 8 target of the SDK the actor runtime serializes the response object using the base response type as declared on the operation on the actor interface. 

### Solution
The programmer must decorate the base response class that is declared on the operation on the actor interface with a `JsonDerivedType` attribute specifying the derived response class. 
Alternatively, a custom `IJsonTypeInfoResolver` implementation can be set on the `JsonSerializerOptions.TypeInfoResolver` property adding a `JsonDerivedType` instance to the `JsonPolymorphismOptions.DerivedTypes` property on the `JsonTypeInfo.PolymorphismOptions` property.

Note that while .NET 7 introduced polymorphic (de)serialization using `JsonDerivedType` attribute or `JsonPolymorphismOptions` on a custom `IJsonTypeInfoResolver` implementation, Dapr does not ship a .NET 7 target of the SDK. Hence, a .NET 7 consumer application uses the .NET 6 target of the Dapr SDK. Therefore, this breaking change only applies when upgrading a consumer application to .NET 8.

Note that the new serialization behavior in the .NET 8 version of the SDK allows for polymorphic deserialization of the response by a weakly-typed actor client. This requires that the programmer specifies a `typeDiscriminator` on the `JsonDerivedType` attribute.